### PR TITLE
Rune init

### PR DESCRIPTION
**Pull request by cmcahoon01**
Power Rune **Pipeline**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,8 +4,9 @@ venv/
...
 .vscode/
 
-*.jpgs
+*.png
+*.jpg
 *.mov
 *.mp4
 *.pyc
-
+*.jpgs
```
Comment (pedramamani) [inline .gitignore]: What is this? Remove.

**Review by pedramamani — CHANGES_REQUESTED** [review]

**Review by pedramamani — APPROVED** [review]

File: source/rune/__main__.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        prediction = rune.process(frame)
+        display_target(frame, prediction)
+        if cv2.waitKey(1) & 0xFF == ord('q'):  # press q to quit
+            break
+        ret, frame = cap.read()  # get next frame
+    cap.release()
+    cv2.destroyAllWindows()
+
+
+def run_webcam(source):
```
Comment (jamesrosstwo) [inline source/rune/__main__.py]: Let's either complete or delete this

File: source/rune/predict_target/assign_panels/assign_panels.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+import matplotlib.pyplot as plt
+from scipy.signal import find_peaks
+from source.common.module import Module
+
+
+class AssignPanels(Module):
+    def __init__(self, state=None):
+        self.working_dir = Path(os.path.dirname(os.path.abspath(__file__)))
+        super().__init__(self.working_dir, state=state)
+        # todo: obtain and set frame size here
```
Comment (jamesrosstwo) [inline source/rune/predict_target/assign_panels/assign_panels.py]: Let's complete this todo before merging

File: source/rune/predict_target/assign_panels/assign_panels.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        if self.properties["mode"] == 'debug':
+            self.plot_lumins(lumins, states)
+            cv2.imshow('Frame B&W', frame_bw)
+            cv2.imshow('Frame BGR', self.frame)
+            cv2.waitKey(1)
+
+        return states
+
+    def calc_lumin_square(self, frame_bw, coords_center):
+        # todo: move size_frame attribute to predict_target.settings?
```
Comment (jamesrosstwo) [inline source/rune/predict_target/assign_panels/assign_panels.py]: image

File: source/rune/predict_target/assign_panels/assign_panels.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        for y in range(round(coords_center[1] - size_square / 2), round(coords_center[1] + size_square / 2)):
+            for x in range(round(coords_center[0] - size_square / 2), round(coords_center[0] + size_square / 2)):
+                cnt_pixels += 1
+                if frame_bw[y][x]:
+                    cnt_whites += 1
+
+        return 255 * cnt_whites/cnt_pixels
+
+    def calc_lumins_squares(self, frame_bw):
+        # todo: move radius to parent module maybe? depends on how he does it.
```
Comment (jamesrosstwo) [inline source/rune/predict_target/assign_panels/assign_panels.py]: image

File: source/rune/predict_target/assign_panels/assign_panels.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            distance=self.properties["angle_peak_distance"]
+        )
+        peak_angles = []
+        for peak_index in peak_indices:
+            peak_angles += [(peak_index-inc_peak_distance+1) % len(lumins) * self.properties["angle_increment_mask"]]
+
+        return peak_angles
+
+    def assign_state(self, lumin):
+        # todo: fix the panel state defined by parent issue
```
Comment (jamesrosstwo) [inline source/rune/predict_target/assign_panels/assign_panels.py]: image

File: source/rune/predict_target/predict_target.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+class PredictTarget(Module):
+    def __init__(self, state=None):
+        self.working_dir = Path(os.path.dirname(os.path.abspath(__file__)))
+        super().__init__(self.working_dir, state=state)
+        self.assign_panels = AssignPanels({"mode": "not debug"})
+        self.frames_calculated = 0
+        self.angles = []
+
+    def process(self, image):
```
Comment (jamesrosstwo) [inline source/rune/predict_target/predict_target.py:18]: Needs docstring

File: source/rune/preprocess/preprocess.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+from source.common.module import Module
+
+
+class Preprocess(Module):
+    def __init__(self, state=None):
+        self.working_dir = Path(os.path.dirname(os.path.abspath(__file__)))
+        super().__init__(self.working_dir, state=state)
+
+    def process(self, frame, rune_center):
```
Comment (jamesrosstwo) [inline source/rune/preprocess/preprocess.py]: Docstring please! It's unclear to me what `rune_center` is. I'm guessing that it's the coordinates of the center? Not really sure.

File: source/rune/preprocess/preprocess.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import os
+from pathlib import Path
+
+from source.common.module import Module
+
+
+class Preprocess(Module):
```
Comment (jamesrosstwo) [inline source/rune/preprocess/preprocess.py]: This name makes the purpose unclear. Would someone looking at this class used in another file immediately be able to understand what `Preprocess` does?

Maybe a name specific to the power rune would be more specific. Something like 
- `PreprocessRune`
- `RunePreprocessor`
- `RunePreprocess`

File: source/rune/preprocess/settings.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
+  "scaling": 0.6
+}
```
Comment (jamesrosstwo) [inline source/rune/preprocess/settings.json]: Newline please!

File: source/rune/rune.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+class Rune(Module):
+    def __init__(self, state=None):
+        self.working_dir = Path(os.path.dirname(os.path.abspath(__file__)))
+        super().__init__(self.working_dir, state=state)
+        self.preprocess = Preprocess()
+        self.predict_target = PredictTarget()
+        self.rune_center = self.properties["rune_center"]
+
+    def process(self, frame):
```
Comment (jamesrosstwo) [inline source/rune/rune.py:16]: Docstring please! Let's make sure we're consistent with documentation.

File: source/rune/settings.json [hunk 12]
```diff
@@ -0,0 +1,6 @@
+{
+  "rune_center": [
+    0.53,
+    0.42
+  ]
+}
```
Comment (jamesrosstwo) [inline source/rune/settings.json:6]: Newline please

**Review by jamesrosstwo — DISMISSED** [review]

**Pull request merged**